<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Abhilash Kannan" />


<title>Computing P-value for Hierarchical Clustering</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Clustering analyis for Candida albicans</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="about.html">
    <span class="fa fa-info"></span>
     
    About
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    More
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Clust_tend.html">Clustering tendency</a>
    </li>
    <li>
      <a href="optimal_clust_identification.html">Optimal number of clusters</a>
    </li>
    <li>
      <a href="valid_stats.html">Validation Statistics</a>
    </li>
    <li>
      <a href="Clustering_p_value.html">P-value for Heirarchial Clustering</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="mailto:abhilifesicizurich@gmail.com">
    <span class="fa fa-question fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Computing P-value for Hierarchical Clustering</h1>
<h4 class="author">Abhilash Kannan</h4>
<h4 class="date">3/4/2020</h4>

</div>


<style type="text/css">

body{ /* Normal  */
  
      
  }
td {  /* Table  */
  
}
h1.title {
  
  color: DarkRed;
}
h1 { /* Header 1 */
  
  color: DarkBlue;
}
h2 { /* Header 2 */
    
  color: DarkBlue;
}
h3 { /* Header 3 */
  
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}

h4 { /* Header 4 */
  
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}

h5 { /* Header 4 */
  
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}


code.r{ /* Code block */
    
}
pre { /* Code block - determines code spacing between lines */
    
}
</style>
<p>Clusters can be found in a data set by chance due to clustering noise or sampling error. This Section describes the R package <strong>pvclust</strong> (Suzuki and Shimodaira 2015) which uses <strong>bootstrap resampling techniques</strong> to compute <em>p-value</em> for each <strong>hierarchical clusters</strong>.</p>
<div id="algorithm" class="section level2">
<h2>2) Algorithm</h2>
<ul>
<li><ol style="list-style-type: decimal">
<li>Generated thousands of bootstrap samples by randomly sampling elements of the data</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Compute hierarchical clustering on each bootstrap copy</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>For each cluster:</li>
</ol></li>
<li><p>compute the bootstrap probability (BP) value which corresponds to the frequency that the cluster is identified in bootstrap copies.</p></li>
<li><p>Compute the approximately unbiased (AU) probability values (p-values) by multiscale bootstrap resampling</p></li>
</ul>
<p><strong>Clusters with AU &gt;= 95% are considered to be strongly supported by data</strong></p>
<div id="install-the-required-packages" class="section level3">
<h3>Install the required packages</h3>
<pre class="r"><code>#Install pvclust:
#install.packages(&quot;pvclust&quot;)
#Load pvclust:
library(pvclust)</code></pre>
</div>
</div>
<div id="data-preparation" class="section level2">
<h2>3) Data preparation</h2>
<p>We’ll use two data sets:</p>
<ul>
<li><p><strong>Gene expression data - df1</strong> (log2FC of isolates vs SC5314 - 182 samples X 6217 genes from Monocultre and Coculture).</p></li>
<li><p><strong>Gene expression data - df2</strong> (log2FC of isolates in Monoculture vs Coculture).</p></li>
</ul>
<pre class="r"><code>set.seed(123)

#df1
logFC_isolatesvsSC5314 &lt;- read.delim(&quot;/Users/Abhi/Desktop/Abhilash/study/Post-Doc/Lab_notebook/eQTL/Gene_expression_analysis/IsolatesvsSC5314/isolatesvsSC5314_LogFC_transpose.txt&quot;,
                          row.names=&quot;isolate&quot;,stringsAsFactors = FALSE)

#df2
logFC_Monovscoculture &lt;- read.delim(&quot;/Users/Abhi/Desktop/Abhilash/study/Post-Doc/Lab_notebook/eQTL/Gene_expression_analysis/Isolates_conditions/Mono_coculture_isolates_LogFC_transpose.txt&quot;,
                          row.names=&quot;isolate_comp&quot;,stringsAsFactors = FALSE)</code></pre>
<p>The data sets look like this:</p>
<pre class="r"><code>#df1
head(logFC_isolatesvsSC5314, 3)

#df2
head(logFC_Monovscoculture, 3)</code></pre>
<p>Check if there is any missing values, if found try to remove it.</p>
<pre class="r"><code>#df1
df1 &lt;- logFC_isolatesvsSC5314[,c(9:6225)]
df1 &lt;- na.omit(df1)
#df2
df2 &lt;- logFC_Monovscoculture[,c(7:6223)]
df2 &lt;- na.omit(df2)</code></pre>
<p>we also don’t want the clustering algorithm to depend to an arbitrary variable unit, we start by scaling/standardizing the data using the R function scale():</p>
<pre class="r"><code>#df1
df1_scaled &lt;- scale(df1)

#df2
df2_scaled &lt;- scale(df2)</code></pre>
<p><strong>Defining Groups df1 and df2 dataset</strong>:</p>
<p>This is required for the cluster validation</p>
<pre class="r"><code># df1

#Groups for df1 based on SNP clades
group_df1_SNP &lt;- factor(logFC_isolatesvsSC5314$CLADE_SNP_numeric, levels = unique(logFC_isolatesvsSC5314$CLADE_SNP_numeric))
group_df1_SNP

#Groups for df1 based on MLST clades
group_df1_MLST &lt;- factor(logFC_isolatesvsSC5314$CLADE_MLST_numeric, levels = unique(logFC_isolatesvsSC5314$CLADE_MLST_numeric))
group_df1_MLST

#Groups for df1 based on site of infection
group_df1_site &lt;- factor(logFC_isolatesvsSC5314$Site_numeric, levels = unique(logFC_isolatesvsSC5314$Site_numeric))
group_df1_site

#Groups for df1 based on culture type
group_df1_culture &lt;- factor(logFC_isolatesvsSC5314$culture_type_numeric, levels = unique(logFC_isolatesvsSC5314$culture_type_numeric))
group_df1_culture

#df2

#Groups for df2 based on SNP clades
group_df2_SNP &lt;- factor(logFC_Monovscoculture$CLADE_SNP_numeric, levels = unique(logFC_Monovscoculture$CLADE_SNP_numeric))
group_df2_SNP

#Groups for df2 based on MLST clades
group_df2_MLST &lt;- factor(logFC_Monovscoculture$CLADE_MLST_numeric, levels = unique(logFC_Monovscoculture$CLADE_MLST_numeric))
group_df2_MLST

#Groups for df2 based on site of infection
group_df2_site &lt;- factor(logFC_Monovscoculture$Site_Numeric, levels = unique(logFC_Monovscoculture$Site_Numeric))
group_df2_site</code></pre>
</div>
<div id="compute-p-value-for-hierarchical-clustering" class="section level2">
<h2>4) Compute p-value for hierarchical clustering</h2>
<div id="description-of-pvclust-function" class="section level3">
<h3>Description of pvclust() function</h3>
<p>The function <em>pvclust()</em> can be used as follow:</p>
<p><strong>pvclust(data, method.hclust = “average”, method.dist = “correlation”, nboot = 1000)</strong></p>
<p>Note that, the computation time can be strongly decreased using parallel computation version called <strong>parPvclust()</strong>. (Read ?parPvclust() for more information.)</p>
<p><strong>parPvclust(cl=NULL, data, method.hclust = “average”,method.dist = “correlation”, nboot = 1000,iseed = NULL)</strong></p>
<ul>
<li><p><strong>data</strong>: numeric data matrix or data frame.</p></li>
<li><p><strong>method.hclust</strong>: the agglomerative method used in hierarchical clustering. Possible values are one of “average”, “ward”, “single”, “complete”, “mcquitty”, “median” or “centroid”. The default is “average”. See method argument in ?hclust.</p></li>
<li><p><strong>method.dist</strong>: the distance measure to be used. Possible values are one of “correlation”, “uncentered”, “abscor” or those which are allowed for method argument in <strong>dist()</strong> function, such “euclidean” and “manhattan”.</p></li>
<li><p><strong>nboot</strong>: the number of bootstrap replications. The default is 1000.</p></li>
<li><p><strong>iseed</strong>: an integrer for random seeds. Use iseed argument to achieve reproducible results.</p></li>
</ul>
<p>The function <em>pvclust()</em> returns an object of class pvclust containing many elements including hclust which contains hierarchical clustering result for the original data generated by the function <em>hclust()</em>.</p>
</div>
<div id="usage-of-pvclust-function" class="section level3">
<h3>Usage of pvclust() function</h3>
<p><em>pvclust()</em> performs clustering on the columns of the data set, which correspond to samples in our case.</p>
<p>The R code below computes <em>pvclust()</em> using 10 as the number of bootstrap replications (for speed):</p>
<p><strong>For df1 dataset</strong>:</p>
<pre class="r"><code>library(pvclust)
set.seed(123)
res.pv_df1 &lt;- pvclust(t(df1_scaled), method.dist=&quot;cor&quot;, 
                  method.hclust=&quot;ward.D2&quot;, nboot = 500)
# Default plot
plot(res.pv_df1, hang = -1, cex = 0.5)
pvrect(res.pv_df1)</code></pre>
<p><img src="Clustering_p_value_files/figure-html/pvclust%20for%20df1%20dataset-1.png" width="672" /></p>
<p><strong>For df1 dataset</strong>:</p>
<pre class="r"><code>library(pvclust)
set.seed(123)
res.pv_df2 &lt;- pvclust(t(df2_scaled), method.dist=&quot;cor&quot;, 
                  method.hclust=&quot;ward.D2&quot;, nboot = 500)
# Default plot
plot(res.pv_df2, hang = -1, cex = 0.5)
pvrect(res.pv_df2)</code></pre>
<p><img src="Clustering_p_value_files/figure-html/pvclust%20for%20df2%20dataset-1.png" width="672" /></p>
<p>To extract the objects from the significant clusters, use the function pvpick():</p>
<pre class="r"><code>clusters_df1 &lt;- pvpick(res.pv_df1)
clusters_df1

clusters_df2 &lt;- pvpick(res.pv_df2)
clusters_df2</code></pre>
<!-- #For df1: -->
<!-- ```{r Parallel computation for df1 dataset, echo=TRUE, results='hide', message=F, warning=F} -->
<!-- # Create a parallel socket cluster -->
<!-- library(parallel) -->
<!-- cl_f1 <- makeCluster(2, type = "PSOCK") -->
<!-- # parallel version of pvclust -->
<!-- res.pv_df1_par <- parPvclust(cl_f1, t(df1_scaled), nboot=1000) -->
<!-- stopCluster(cl_f1) -->
<!-- ``` -->
<!-- #For df1: -->
<!-- ```{r Parallel computation for df2 dataset, echo=TRUE, results='hide', message=F, warning=F} -->
<!-- # Create a parallel socket cluster -->
<!-- library(parallel) -->
<!-- cl_f2 <- makeCluster(2, type = "PSOCK") -->
<!-- # parallel version of pvclust -->
<!-- res.pv_df2_par <- parPvclust(cl_f2, t(df2_scaled), nboot=1000) -->
<!-- stopCluster(cl_f2) -->
<!-- ``` -->
</div>
</div>

<p><b>Copyright &copy; 2020 Abhilash Kannan.</b></p>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
